File: shieldai/api/src/services/languageDetector.ts
/**
 * Language detector — uses script analysis and word frequency heuristics.
 * For production, this delegates to the ML sidecar (fasttext lid.176.bin).
 * Falls back to local heuristic detection if ML sidecar is unavailable.
 */

import { detectScript } from '../utils/transliteration';
import { getEnv } from '../config/env';
import { logger } from '../config/logger';

export interface LanguageDetectionResult {
  language: string;          // ISO 639-1: en, hi, ta, te, bn, hi-en, etc.
  script: string;            // latin, devanagari, tamil, telugu, bengali, mixed
  confidence: number;        // 0.0 - 1.0
  isCodeMixed: boolean;
  detectedVia: 'ml_sidecar' | 'local_heuristic';
}

// Hindi word frequency list (romanized)
const HINDI_MARKERS = /\b(kya|hai|hain|hoon|mein|mujhe|tumhare|aapka|nahi|nahin|bhi|aur|lekin|sabhi|saare|kripya|namaste|bahut|accha|pehle|baad|yahan|wahan|bhai|yaar|dekh|sun|bol|chal|kar|karo|karna|ho|hota|hoti|raha|rahi|rahe|tha|thi|the|wala|wali|wale|dena|lena|jana|aana)\b/gi;

const TAMIL_MARKERS = /\b(enna|epdi|yaar|inge|ange|nalla|romba|konjam|sollu|vaanga|ponga|irukku|illai|theriyum|ungal|enakku|ithu|athu|intha|antha|vera|mattum|thaan)\b/gi;

const TELUGU_MARKERS = /\b(emi|ela|evaru|ikkada|akkada|manchi|chala|koncham|cheppu|undi|ledu|teliyadu|meeru|naaku|vaallu|idi|adi|eedhi|adhi|mari|aithe|kaani)\b/gi;

const BENGALI_MARKERS = /\b(ki|keno|kothay|ekhane|okhane|bhalo|boro|chhoto|onek|ektu|bolo|aasho|jao|aacho|nei|jani|aapnar|aamaar|ora|eta|ota)\b/gi;

/**
 * Attempt ML sidecar language detection via HTTP.
 */
async function detectViaMlSidecar(text: string): Promise<LanguageDetectionResult | null> {
  try {
    const env = getEnv();
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), env.ML_SERVICE_TIMEOUT_MS);

    const response = await fetch(`${env.ML_SERVICE_URL}/detect-language`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ text }),
      signal: controller.signal,
    });

    clearTimeout(timeout);

    if (!response.ok) return null;

    const data = await response.json() as {
      language: string;
      confidence: number;
      is_code_mixed: boolean;
    };

    const script = detectScript(text);
    return {
      language: data.is_code_mixed ? `${data.language}-en` : data.language,
      script,
      confidence: data.confidence,
      isCodeMixed: data.is_code_mixed,
      detectedVia: 'ml_sidecar',
    };
  } catch (err) {
    logger.debug({ err }, 'ML sidecar language detection unavailable, using local heuristic');
    return null;
  }
}

/**
 * Local heuristic language detection — counts word matches per language.
 */
function detectLocal(text: string): LanguageDetectionResult {
  const script = detectScript(text);

  // Native scripts are definitive
  if (script === 'devanagari') return { language: 'hi', script, confidence: 0.95, isCodeMixed: false, detectedVia: 'local_heuristic' };
  if (script === 'tamil') return { language: 'ta', script, confidence: 0.95, isCodeMixed: false, detectedVia: 'local_heuristic' };
  if (script === 'telugu') return { language: 'te', script, confidence: 0.95, isCodeMixed: false, detectedVia: 'local_heuristic' };
  if (script === 'bengali') return { language: 'bn', script, confidence: 0.95, isCodeMixed: false, detectedVia: 'local_heuristic' };

  // Mixed script
  if (script === 'mixed') {
    return { language: 'hi-en', script, confidence: 0.80, isCodeMixed: true, detectedVia: 'local_heuristic' };
  }

  // Latin script — count word markers
  const counts: Record<string, number> = {
    hi: (text.match(HINDI_MARKERS) || []).length,
    ta: (text.match(TAMIL_MARKERS) || []).length,
    te: (text.match(TELUGU_MARKERS) || []).length,
    bn: (text.match(BENGALI_MARKERS) || []).length,
  };

  const totalIndic = Object.values(counts).reduce((a, b) => a + b, 0);
  const words = text.split(/\s+/).length;

  if (totalIndic < 2 || totalIndic / words < 0.1) {
    return { language: 'en', script: 'latin', confidence: 0.85, isCodeMixed: false, detectedVia: 'local_heuristic' };
  }

  const topLang = Object.entries(counts).sort((a, b) => b[1] - a[1])[0];
  const indicRatio = totalIndic / words;
  const isCodeMixed = indicRatio > 0.1 && indicRatio < 0.7;

  return {
    language: isCodeMixed ? `${topLang[0]}-en` : topLang[0],
    script: 'latin',
    confidence: Math.min(0.9, 0.5 + indicRatio),
    isCodeMixed,
    detectedVia: 'local_heuristic',
  };
}

/**
 * Detect language — tries ML sidecar first, falls back to local heuristic.
 */
export async function detectLanguage(text: string): Promise<LanguageDetectionResult> {
  const mlResult = await detectViaMlSidecar(text);
  if (mlResult) return mlResult;
  return detectLocal(text);
}

/**
 * Synchronous local-only detection (for use in preprocessor).
 */
export function detectLanguageLocal(text: string): LanguageDetectionResult {
  return detectLocal(text);
}

File: shieldai/ml-service/app/models/language_detector.py
"""
Language detector using fasttext lid.176.bin model.
Detects 176 languages including all major Indian languages.
Falls back to heuristic detection if fasttext is unavailable.
"""

import re
import time
import logging
from typing import Optional

logger = logging.getLogger(__name__)

_model = None
_is_loaded = False

# Indian language codes supported
INDIC_LANGUAGES = {"hi", "ta", "te", "bn", "mr", "gu", "kn", "ml", "pa", "or", "as", "ur"}

# Hindi romanized word patterns for code-mix detection
HINDI_MARKERS = re.compile(
    r"\b(kya|hai|hain|hoon|mein|mujhe|tumhare|aapka|nahi|nahin|bhi|aur|lekin|"
    r"sabhi|saare|kripya|namaste|bahut|accha|pehle|baad|yahan|wahan|bhai|yaar|"
    r"dekh|sun|bol|chal|kar|karo|karna|ho|hota|hoti|raha|rahi|rahe|tha|thi|the|"
    r"wala|wali|wale|dena|lena|jana|aana|apna|tera|tumhara)\b",
    re.IGNORECASE,
)

# Script detection regex
DEVANAGARI = re.compile(r"[\u0900-\u097F]")
TAMIL_SCRIPT = re.compile(r"[\u0B80-\u0BFF]")
TELUGU_SCRIPT = re.compile(r"[\u0C00-\u0C7F]")
BENGALI_SCRIPT = re.compile(r"[\u0980-\u09FF]")


def load_model(cache_dir: str = "/app/models"):
    """Load fasttext language identification model."""
    global _model, _is_loaded
    try:
        import fasttext

        model_path = f"{cache_dir}/lid.176.bin"
        try:
            _model = fasttext.load_model(model_path)
            _is_loaded = True
            logger.info(f"Loaded fasttext language model from {model_path}")
        except Exception:
            # Try downloading via fasttext
            logger.warning(f"Model not found at {model_path}, downloading...")
            import urllib.request
            url = "https://dl.fbaipublicfiles.com/fasttext/supervised-models/lid.176.bin"
            urllib.request.urlretrieve(url, model_path)
            _model = fasttext.load_model(model_path)
            _is_loaded = True
            logger.info("Downloaded and loaded fasttext language model")
    except Exception as e:
        logger.error(f"Failed to load fasttext model: {e}. Using heuristic detection.")
        _is_loaded = False


def is_loaded() -> bool:
    return _is_loaded


def detect(text: str) -> dict:
    """
    Detect language of input text.
    Returns language code, confidence, and code-mixing flag.
    """
    start = time.time()

    # Clean text for detection
    clean_text = text.replace("\n", " ").strip()
    if not clean_text:
        return _result("en", 1.0, False, start)

    # Check native scripts first (definitive)
    script_lang = _detect_by_script(clean_text)
    if script_lang:
        is_mixed = bool(re.search(r"[a-zA-Z]{3,}", clean_text))
        return _result(script_lang, 0.95, is_mixed, start)

    # Try fasttext
    if _is_loaded and _model is not None:
        try:
            predictions = _model.predict(clean_text, k=3)
            labels = [l.replace("__label__", "") for l in predictions[0]]
            scores = predictions[1].tolist()

            primary_lang = labels[0]
            primary_score = scores[0]

            # Check for code-mixing: if top-2 languages differ significantly
            is_mixed = False
            if len(labels) >= 2 and scores[1] > 0.15:
                lang_set = set(labels[:2])
                if lang_set & INDIC_LANGUAGES and "en" in lang_set:
                    is_mixed = True
                    primary_lang = [l for l in labels[:2] if l in INDIC_LANGUAGES][0]

            # Also check for romanized Hindi words in Latin text
            if primary_lang == "en" and _has_hindi_words(clean_text):
                is_mixed = True
                primary_lang = "hi"

            return _result(primary_lang, primary_score, is_mixed, start)
        except Exception as e:
            logger.error(f"Fasttext detection error: {e}")

    # Heuristic fallback
    return _detect_heuristic(clean_text, start)


def _detect_by_script(text: str) -> Optional[str]:
    """Detect language from script (Devanagari, Tamil, etc.)."""
    if DEVANAGARI.search(text):
        return "hi"
    if TAMIL_SCRIPT.search(text):
        return "ta"
    if TELUGU_SCRIPT.search(text):
        return "te"
    if BENGALI_SCRIPT.search(text):
        return "bn"
    return None


def _has_hindi_words(text: str) -> bool:
    """Check if Latin-script text contains romanized Hindi words."""
    matches = HINDI_MARKERS.findall(text)
    words = text.split()
    return len(matches) >= 2 and len(matches) / max(len(words), 1) > 0.1


def _detect_heuristic(text: str, start: float) -> dict:
    """Simple heuristic language detection for when fasttext is unavailable."""
    if _has_hindi_words(text):
        # Check if it's code-mixed
        english_words = len(re.findall(r"\b[a-zA-Z]{4,}\b", text))
        total_words = len(text.split())
        is_mixed = english_words / max(total_words, 1) > 0.2
        return _result("hi", 0.6, is_mixed, start)
    return _result("en", 0.7, False, start)


def _result(language: str, confidence: float, is_code_mixed: bool, start: float) -> dict:
    elapsed = (time.time() - start) * 1000
    return {
        "language": language,
        "confidence": round(confidence, 4),
        "is_code_mixed": is_code_mixed,
        "time_ms": round(elapsed, 2),
    }
